runner_test.go

```diff
@@ -367,6 +357,250 @@ func (*RunnerSuite) TestFatalErrorWhileSelfStartWorker(c *gc.C) {
 	c.Assert(err, gc.Equals, fatalStarter.startErr)
 }
 
+func (*RunnerSuite) TestWorkerWithNoWorker(c *gc.C) {
+	runner := worker.NewRunner(worker.RunnerParams{
+		IsFatal:      allFatal,
+		RestartDelay: time.Millisecond,
+	})
+	w, err := runner.Worker("id", nil)
+	c.Assert(errors.Cause(err), gc.Equals, worker.ErrNotFound)
+	c.Assert(w, gc.Equals, nil)
+}
+
+func (*RunnerSuite) TestWorkerWithWorkerImmediatelyAvailable(c *gc.C) {
+	runner := worker.NewRunner(worker.RunnerParams{
+		IsFatal:      allFatal,
+		RestartDelay: time.Millisecond,
+	})
+	starter := newTestWorkerStarter()
+	runner.StartWorker("id", starter.start)
+
+	// Wait long enough to be reasonably confident that the
+	// worker has been started and registered.
+	time.Sleep(10 * time.Millisecond)
```

Comment:
The result is the same whether we sleep for a long time or not - there's no externally visible difference and the test will pass either way.

We could insert test synchronisation points into the production code, I guess, but I'd prefer not.

If occasionally the code takes longer than 10ms to add the worker to the map, we'll just go through the slow path (the path that TestWorkerWithWorkerNotImmediatelyAvailable tests).